### PR TITLE
SCI-F template

### PR DESCRIPTION
**Pull request by bilke**
WIP: Just want to get some feedback if a template is proper structure for this?

SCI-F, the Scientific Filesystem:

- https://sci-f.github.io
- https://academic.oup.com/gigascience/article/7/5/giy023/4931737

Usage:

```python
Stage0 += shell(commands=['python3 -m pip install scif'])
Stage0 += runscript(commands=["scif \"$@\""])

scif_mpi = scif(
  name='mpi-bandwidth',
  install=['wget -q -nc --no-check-certificate -P /var/tmp https://computing.llnl.gov/tutorials/mpi/samples/C/mpi_bandwidth.c',
           'mpicc -o bin/mpi-bandwidth /var/tmp/mpi_bandwidth.c'],
  run='exec /scif/mpi-bandwidth/bin/mpi-bandwidth "$@"',
  help='This app provides a MPI bandwidth test program',
  test='exec mpirun -np 2 /scif/apps/mpi-bandwidth/bin/mpi-bandwidth "$@"'
)
Stage0 += scif_mpi.install()
```

When you have built the container you can run scif apps by name:
```bash
$ singularity run scif.simg apps
mpi-bandwidth
[other apps]

$ mpirun -np 2 singularity run scif.simg run mpi-bandwidth
```

### CODE REVIEW

**Comment by samcmill** [comment]
The closest parallels to SCIF seem to be EasyBuild and Spack.  In those cases, there are HPCCM recipes (the would also be appropriate as building blocks).  The recipes take an input, e.g., an EasyConfig and build the corresponding container image.  The EasyConfig itself is defined outside HPCCM.

SCIF could be implemented similarly.  A HPCCM SCIF recipe that installs the SCIF environment, and if a SCIF recipe is provided, continues on to run SCIF.

Another approach would be to add a attribute to some of the existing primitives (and perhaps create some new primitives?)  Something like:

```
Stage0 += environment(variables={'PATH': '/mypath'}, _app='foo')
```

For Singularity, this would generate:

```
%appenv foo
   export PATH=/mypath
```

Without the `_app='foo'` attribute, this would generate the current `%environment` block.


Note `_` has been used to denote that options that are specific to a particular container specification format and not general.

**Comment by bilke** [comment]
Thanks for the explanations! I really like the approach to add attributes to existing primitives and will start re-implementing it!

**Comment by bilke** [comment]
@samcmill I refactored this and also the runscript implementation. What do you think?

File: hpccm/primitives/copy.py [hunk 1]
```diff
@@ -38,11 +38,18 @@ def __init__(self, **kwargs):
         self.__dest = kwargs.get('dest', '')
         self.__from = kwargs.get('_from', '') # Docker specific
         self.__src = kwargs.get('src', '')
+        self._app = kwargs.get('_app', '') # Singularity specific
```
Comment (samcmill) [inline hpccm/primitives/copy.py]: Please order these alphabetically

File: hpccm/primitives/copy.py [hunk 2]
```diff
@@ -38,11 +38,18 @@ def __init__(self, **kwargs):
...
 
     def __str__(self):
         """String representation of the primitive"""
         if self.__dest and self.__src:
             if hpccm.config.g_ctype == container_type.DOCKER:
+                if self._app:
```
Comment (samcmill) [inline hpccm/primitives/copy.py]: I think Docker should still output instructions in this case?  Behave as if the `_app` attribute was not set.

File: hpccm/primitives/runscript.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+                                    'equivalent: ignoring statement!')
+                    return ''
+                if len(self.commands) > 1:
+                    logging.warning('Multiple commands given to runscript.'
+                                    'Docker ENTRYPOINT supports just one cmd:'
+                                    'ignoring remaining commands!')
+                # Format:
+                # ENTRYPOINT ["cmd1", "arg1", "arg2", ...]
+                s = []
+                s.extend('"{}"'.format(x) for x in self.commands[0].split(" "))
```
Comment (samcmill) [inline hpccm/primitives/runscript.py]: Use the shlex module for splitting and quoting?  Use six.moves for shlex_quote for portability between Python 2 and 3.

File: hpccm/primitives/runscript.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        """String representation of the primitive"""
+        if self.commands:
+            if hpccm.config.g_ctype == container_type.DOCKER:
+                if self._app:
+                    logging.warning('The Singularity specific %app.. syntax was'
+                                    'requested. Docker does not have an '
+                                    'equivalent: ignoring statement!')
+                    return ''
+                if len(self.commands) > 1:
+                    logging.warning('Multiple commands given to runscript.'
```
Comment (samcmill) [inline hpccm/primitives/runscript.py]: Spaces between sentences.

File: hpccm/primitives/runscript.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+                # Format:
+                # ENTRYPOINT ["cmd1", "arg1", "arg2", ...]
+                s = []
+                s.extend('"{}"'.format(x) for x in self.commands[0].split(" "))
+                return 'ENTRYPOINT [' + ', '.join(s) + ']'
+
+            elif hpccm.config.g_ctype == container_type.SINGULARITY:
+                # prepend last command with exec
+                last_cmd_index = len(self.commands) - 1
+                self.commands[last_cmd_index] = 'exec {0}'.format(self.commands[last_cmd_index])
```
Comment (samcmill) [inline hpccm/primitives/runscript.py]: self.commands[-1] = ...

File: hpccm/primitives/shell.py [hunk 6]
```diff
@@ -50,17 +57,20 @@ def __str__(self):
...
-                    s.append('    cd /')
+                if self._app:
+                    s = ['%appinstall {0}'.format(self._app)]
+                else:
+                    s = ['%post']
+                    # For consistency with Docker. Docker resets the
+                    # working directory to '/' at the beginning of each
+                    # 'RUN' instruction. Singularity %appinstall is already
+                    # run in its own working directory.
+                    if self.chdir:
```
Comment (samcmill) [inline hpccm/primitives/shell.py]: un-indent?

Comment (bilke) [reply hpccm/primitives/shell.py]: No, `%appinstall` sets the current working directory to the app base dir (`/scif/apps/[name]`). This behavior should be preserved in my opinion.

Comment (samcmill) [reply hpccm/primitives/shell.py]: Make sense.  Can you please add a comment saying this?

**Comment by samcmill** [comment]
> I solved it in my MPI test app by sourcing some Singularity env files:

Clever.  I wonder if it would be useful to have a parameter to source the environment scripts?

Something like:

```
Stage0 += shell(commands=[...], _app=app, _appenv=True)
```

The default would be False, and does nothing unless `_app` is specified.

File: hpccm/primitives/runscript.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+    def __str__(self):
+        """String representation of the primitive"""
+        if self.commands:
+            if hpccm.config.g_ctype == container_type.DOCKER:
+                if self._app:
+                    logging.warning('The Singularity specific %app.. syntax was'
+                                    'requested. Docker does not have an '
+                                    'equivalent: ignoring statement!')
+                    return ''
```
Comment (samcmill) [inline hpccm/primitives/runscript.py]: Please remove the empty return

File: hpccm/primitives/runscript.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+        self._app = kwargs.get('_app', '') # Singularity specific
+        self.commands = kwargs.get('commands', [])
+
+    def __str__(self):
+        """String representation of the primitive"""
+        if self.commands:
+            if hpccm.config.g_ctype == container_type.DOCKER:
+                if self._app:
+                    logging.warning('The Singularity specific %app.. syntax was'
+                                    'requested. Docker does not have an '
```
Comment (samcmill) [inline hpccm/primitives/runscript.py]: missing space

**Pull request merged**